engine/common/follower/pending_tree/pending_tree.go

```diff
@@ -0,0 +1,198 @@
+// Internally, the mempool utilizes the LevelledForest.
+// PendingTree is NOT safe to use in concurrent environment.
+// NOTE: PendingTree relies on notion of `CertifiedBlock` which is a valid block which is certified by corresponding QC.
+// This works well for consensus follower as it is designed to work with certified blocks. To use this structure for consensus
+// participant we can abstract out CertifiedBlock or replace it with a generic argument that satisfies some contract(returns View, Height, BlockID).
+// With this change this structure can be used by consensus participant for tracking connection to the finalized state even without
+// having QC but relying on payload validation.
+type PendingTree struct {
+	forest          *forest.LevelledForest
+	lastFinalizedID flow.Identifier
+}
+
+// NewPendingTree creates new instance of PendingTree. Accepts finalized block to set up initial state.
+func NewPendingTree(finalized *flow.Header) *PendingTree {
+	return &PendingTree{
+		forest:          forest.NewLevelledForest(finalized.View),
+		lastFinalizedID: finalized.ID(),
+	}
+}
+
+// AddBlocks accepts a batch of certified blocks, adds them to the tree of pending blocks and finds blocks connected to the finalized state.
+// This function performs processing of incoming certified blocks, implementation is split into a few different sections
+// but tries to be optimal in terms of performance to avoid doing extra work as much as possible.
+// This function follows next implementation:
+//  1. Sorts incoming batch by height. Since blocks can be submitted in random order we need to find blocks with
+//     the lowest height since they are candidates for being connected to the finalized state.
+//  2. Filters out blocks that are already finalized.
+//  3. Deduplicates incoming blocks. We don't store additional vertices in tree if we have that block already stored.
+//  4. Checks for exceeding byzantine threshold. Only one certified block per view is allowed.
+//  5. Finally, blocks with the lowest height from incoming batch that connect to the finalized state we will
+//     mark all descendants as connected, collect them and return as result of invocation.
+//
+// This function is designed to perform resolution of connected blocks(resolved block is the one that connects to the finalized state)
+// using incoming batch. Each block that was connected to the finalized state is reported once.
+// Expected errors during normal operations:
+//   - model.ByzantineThresholdExceededError - detected two certified blocks at the same view
+func (t *PendingTree) AddBlocks(certifiedBlocks []CertifiedBlock) ([]CertifiedBlock, error) {
+	// sort blocks by height, so we can identify if there are candidates for being connected to the finalized state.
+	slices.SortFunc(certifiedBlocks, func(lhs CertifiedBlock, rhs CertifiedBlock) bool {
+		return lhs.Height() < rhs.Height()
+	})
```

Comment:
I am not sure why you are sorting. I think your algorithm is completely order agnostic. Id doesn't care whether you insert a disconnected block first and then connect it later or insert the connector first. 

The sorting consumes CPU, and I am not sure what the benefits are. Wondering what I am missing ..? 🤔

Reply:
I think it's not needed anymore, it was needed for previous implementation. I thought that there is some benefit of processing blocks in their natural order(due to lower number of recursion to collect the fork) but that's very debatable.